pkgs/native_assets_builder/lib/src/utils/file.dart

```diff
@@ -50,4 +51,99 @@ extension DirectoryExtension on Directory {
     }
     return last;
   }
+
+  /// Run [callback] with this Dart process having exclusive access to this
+  /// [Directory].
+  ///
+  /// Note multiple isolates and isolate groups in the same Dart process share
+  /// locks, so these will be able to enter the exclusive section simultanously.
+  ///
+  /// The required [timeout] parameter determines how long to retry before
+  /// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+  // TODO(dacoharkes): Can we also make it in-process safe?
+  // TODO(dacoharkes): Where do we place this code so hook writers can reuse it?
+  // I don't really want to add it to package:native_assets_cli. A new package
+  // in dart-lang/tools?
+  Future<T> exclusive<T>(
+    Future<T> Function() callback, {
+    required Duration timeout,
+  }) async {
+    const lockFileName = '.lock';
+    final lockFile = File.fromUri(uri.resolve(lockFileName));
+    final fileSystemExceptions = <FileSystemException>[];
+    await _ensureLockfileExists(lockFile, fileSystemExceptions);
+
+    final timeoutCompleter = Completer<void>();
+    var timedOut = false;
+    final timeoutTimer = Timer(timeout, () {
+      timedOut = true;
+      timeoutCompleter.complete();
+    });
+
+    final randomAccessFile = await lockFile.open(mode: FileMode.write);
+    await lockFile.writeAsString(
+      'Last acquired by ${Platform.resolvedExecutable} '
+      'running ${Platform.script} on ${DateTime.now()}.',
+    );
+    var ranCallback = false;
+    T? result;
+    final lockAndCallbackFuture =
+        randomAccessFile.lock(FileLock.blockingExclusive).then((lock) async {
```

Comment:
`flutter ...` will print `Waiting to obtain ... lock` when it cannot acquire it immediately. Wouldn't we want to do something similar?